test: make test runner scene ignored by default for BaseMultiInstanceTest

NoelStephensUnity opened the pull request:
This just updates the BaseMultiInstanceTest class to automatically ignore the testrunner scene when clients are synchronizing as well as registers the testrunner scene automatically for NetworkObject synchronization.  The primary purpose behind this PR is to just further simplify BaseMultiInstanceTest usage without having to re-write the same code repeatedly.

NoelStephensUnity committed:
refactor

Specific to testing, this adds the ability to ignore the warning message when scene validation fails so we don't spam log.

NoelStephensUnity committed:
refactor

If not already set, making the test runner scene a default scene for clients to not try to synchronize.  This also will register the test runner scene so that other associated warning messages will not log-spam.

NoelStephensUnity committed:
refactor

This update allows one to add an additional start initialization action to be invoked when just after the server-host is started and after each client is started for scenarios where you need to access sub-systems that are only initialized when the NetworkManager is started (client, host, or server).

Made minor modifications to the BaseMultiInstanceTest to reflect the above changes.

NoelStephensUnity committed:
style

included 1 whitespace

NoelStephensUnity committed:
style

updated comment and removed a LF

NoelStephensUnity committed:
style

updated comment.

NoelStephensUnity committed:
Merge branch 'develop' into test/default-multiInstance-testrunner-scene-validation

SamuelBellomo commented on com.unity.netcode.gameobjects/Tests/Runtime/MultiInstanceHelpers.cs:
```diff
@@ -155,9 +157,21 @@ public static bool Start(bool host, NetworkManager server, NetworkManager[] clie
+            // if set, then invoke this for the server
+            if (startInitializationCallback != null)
+            {
+                startInitializationCallback.Invoke(server);
```
you can do startInitializationCallback?.Invoke(server); with the question mark, without the null check around

NoelStephensUnity replied:
Yep!  Done!

SamuelBellomo commented on com.unity.netcode.gameobjects/Tests/Runtime/BaseMultiInstanceTest.cs:
```diff
@@ -53,6 +56,51 @@ public virtual IEnumerator Teardown()
+
+            // Register the test runner scene so it will be able to synchronize NetworkObjects without logging a
+            // warning about using the currently active scene
+            var scene = SceneManager.GetActiveScene();
```
I'm a bit confused about this sceneManager, which NetworkManager is it associated with?

NoelStephensUnity replied:
This is not a "networkManager.SceneManager" but is a "UnityEngine.SceneManagement.SceneManager".   Just doing a sanity check that the developer writing the integration test hasn't already changed the currently active scene... if so then I don't handle the rest.

SamuelBellomo commented on com.unity.netcode.gameobjects/Tests/Runtime/BaseMultiInstanceTest.cs:
```diff
@@ -53,6 +56,51 @@ public virtual IEnumerator Teardown()
+        private bool VerifySceneIsValidForClientsToLoad(int sceneIndex, string sceneName, LoadSceneMode loadSceneMode)
+        {
+            // exclude test runner scene
+            if (sceneName.StartsWith(k_FirstPartOfTestRunnerSceneName))
```
I'd put at least a debug log somewhere to warn about this. Just in case someone has the brillant idea of trying to use "InitTestScene" as their scene name later.

NoelStephensUnity replied:
I could... but I am hoping that we will migrate all of this to the MultiProcess stuff once that is all sorted.  If it turns out that this becomes a problem with internal Unity developers creating scenes that start with "InitTestScene", then I will do soemthing like that... but the hope is that this won't happen.  :)

SamuelBellomo commented on com.unity.netcode.gameobjects/Tests/Runtime/BaseMultiInstanceTest.cs:
```diff
@@ -53,6 +56,51 @@ public virtual IEnumerator Teardown()
+        }
+
+        /// <summary>
+        /// This registers scene validation with the server to prevent the server from telling connecting
```
I'm a bit confused by your comments.
a few lines above, you say "We want to exclude the TestRunner scene on the host-server side so it won't"
but here you say "This registers scene validation with the server to prevent the server from"
is the test scene included or excluded?

NoelStephensUnity replied:
I will change that comment to:
>         /// This registers scene validation callback for the server to prevent it from telling connecting
>         /// clients to synchronize (i.e. load) the test runner scene.  This will also register the test runner
>         /// scene and its handle for both client(s) and server-host.

But the phrase is correct... this is called for every NetworkManager instance.  If the NetworkManager instance happens to be the server's, then it registers the VerifySceneIsValidForClientsToLoad callback.  

When the server is putting together the synchronization SceneEvent for each client this validation callback will be invoked for every scene the server already has loaded and if the callback returns false then the server will not include that scene as a scene the clients need to "synchronize".   The issue, primarily, is that the temporary Test runner scene isn't in the Scenes in Build list initially and will cause issues if the clients try to load it.
Since the clients and server-host share the same scenes loaded, we also have to include the Test runner scene (since most integration tests using this class don't change the currently active scene) in order to avoid the issue where NetworkObjects are being synchronized to the TestRunner scene.  While this won't crash anything, it will generate a warning and then use the currently active scene.  This just removes that warning to prevent spam as well as any confusion why they are getting the warning in the event a developer is trying to debug an issue.

NoelStephensUnity replied:
If you want, we can huddle briefly tomorrow and I can explain it possibly better that way and you can ask me any additional questions.  It is a little bit confusing at first.

NoelStephensUnity committed:
refactor

updated comment.
using startInitializationCallback?.Invoke(server); as opposed to checking for null.
